File: projects/2026-07-29_github-actions-matrix-secrets/tests/test_app.js
describe('Math operations', () => {
  function add(a, b) {
    return a + b;
  }

  function subtract(a, b) {
    return a - b;
  }

  function multiply(a, b) {
    return a * b;
  }

  function divide(a, b) {
    if (b === 0) {
      throw new Error('Cannot divide by zero');
    }
    return a / b;
  }

  test('add positive numbers', () => {
    expect(add(2, 3)).toBe(5);
  });

  test('add negative numbers', () => {
    expect(add(-1, 1)).toBe(0);
  });

  test('subtract numbers', () => {
    expect(subtract(5, 3)).toBe(2);
  });

  test('multiply numbers', () => {
    expect(multiply(4, 5)).toBe(20);
  });

  test('divide numbers', () => {
    expect(divide(10, 2)).toBe(5);
  });

  test('divide by zero throws error', () => {
    expect(() => divide(10, 0)).toThrow('Cannot divide by zero');
  });
});

describe('Environment variables', () => {
  test('can access environment variables', () => {
    process.env.TEST_VAR = 'test-value';
    expect(process.env.TEST_VAR).toBe('test-value');
  });
});

describe('Matrix support', () => {
  test('verify Node.js version', () => {
    const version = process.versions.node;
    expect(version).toBeDefined();
    console.log(`Running on Node.js ${version}`);
  });
});

File: projects/2026-07-29_github-actions-matrix-secrets/tests/test_app.py
import pytest
from unittest.mock import patch


def add(a, b):
    """Simple addition function for demo"""
    return a + b


def subtract(a, b):
    """Simple subtraction function for demo"""
    return a - b


def multiply(a, b):
    """Simple multiplication function for demo"""
    return a * b


def divide(a, b):
    """Simple division function for demo"""
    if b == 0:
        raise ValueError("Cannot divide by zero")
    return a / b


class TestMath:
    """Math operation tests"""

    def test_add_positive(self):
        assert add(2, 3) == 5

    def test_add_negative(self):
        assert add(-1, 1) == 0

    def test_add_zero(self):
        assert add(0, 0) == 0

    def test_subtract(self):
        assert subtract(5, 3) == 2

    def test_multiply(self):
        assert multiply(4, 5) == 20

    def test_divide(self):
        assert divide(10, 2) == 5.0

    def test_divide_by_zero(self):
        with pytest.raises(ValueError):
            divide(10, 0)


class TestEnvironment:
    """Test environment variable handling"""

    @patch.dict('os.environ', {'API_KEY': 'test-key', 'ENV': 'test'})
    def test_env_variables(self):
        import os
        assert os.environ.get('API_KEY') == 'test-key'
        assert os.environ.get('ENV') == 'test'


def test_matrix_support():
    """Verify tests run on multiple Python versions"""
    import sys
    assert sys.version_info >= (3, 9)


@pytest.fixture
def sample_data():
    """Provide sample test data"""
    return {'value': 42, 'name': 'test'}


def test_with_fixture(sample_data):
    """Test using fixtures"""
    assert sample_data['value'] == 42
    assert sample_data['name'] == 'test'
